### PR TITLE
SystemZ: Handle gr128 to fp128 copies in copyPhysReg

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

File: llvm/test/CodeGen/SystemZ/copy-phys-reg-gr128-to-fp128.mir [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    ; CHECK: liveins: $r0q
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: $f0q = KILL undef $r0q
+    ; CHECK-NEXT: Return implicit $f0q
+    $f0q = COPY undef $r0q
+    Return implicit $f0q
+...
+
+---
+name:            copy_gr128_to_fp128__r0q_to_f0q_subreg0
```
Comment (uweigand) [inline llvm/test/CodeGen/SystemZ/copy-phys-reg-gr128-to-fp128.mir:52]: Unless I'm missing something, this test is completely identical to the `copy_gr128_to_fp128__r0q_to_f0q` one.  How is this supposed to look at just subreg0 here?

File: llvm/test/CodeGen/SystemZ/copy-phys-reg-gr128-to-fp128.mir [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: $f0d = LDGR $r0d, implicit-def $f0q
+    ; CHECK-NEXT: $f2d = LDGR $r1d
+    ; CHECK-NEXT: Return implicit $f0q
+    $f0q = COPY $r0q
+    Return implicit $f0q
+...
+
+---
+name:            copy_gr128_to_fp128__r0q_to_f0q_subreg1
```
Comment (uweigand) [inline llvm/test/CodeGen/SystemZ/copy-phys-reg-gr128-to-fp128.mir:68]: Same here.

**Review by uweigand — APPROVED** [review]
LGTM, but see comment about (apparently) duplicated test cases.

**Comment by arsenm** [comment]
>  I guess you ran into this with an early version of the f128 atomic load/store bitcast patch, but the current version splits into two 64-bit copies very early and never trigger this code path either.

No, I ran into this with the better version (https://github.com/arsenm/llvm-project/tree/systemz-legal-v2i64). v2i64 should be legal, and avoids all of this custom lowering hacking around the SelectionDAG requirement to avoid illegal types. What's pushed now is where I landed after I gave up on making sure none of the ABIs changed 

> 
> Unless I'm missing something here, I'd prefer to revert this PR again rather than keep this as dead code.

copyPhysReg really should be comprehensive. This is tested code so I don't see a compelling reason to remove it and I strongly suggest leaving it. This is avoiding future work for someone 

It's not guaranteed that the DAG never emits these copies. When I was working on the legal type patch, these were slipping in from not-100% precise selection pattern predicates. Having mandatory optimization is bad, and this should always be available as a fallback.

The DAG is also not the only possible source of copies, any machine pass should be able to insert them. GlobalIsel also does away with the "type = register class" DAG reasoning. The fact this was missing also shows there's a missed opportunity for RA to avoid spills by copying to these other classes. This would work if there was a new allocatable synthetic class that was the union of FPR128/GPR128

**Pull request merged**